synapseclient/models/mixins/table_components.py

```diff
@@ -117,6 +141,350 @@ def test_import_pandas() -> None:
+        "concreteType": "org.sagebionetworks.repo.model.table.QueryBundleRequest",
+        "query": {
+            "sql": query,
+            "includeEntityEtag": True,
+        },
+    }
+
+    if part_mask:
+        query_bundle_request["partMask"] = part_mask
+    if limit is not None:
+        query_bundle_request["query"]["limit"] = limit
+    if offset is not None:
+        query_bundle_request["query"]["offset"] = offset
+
+    uri = "/entity/{id}/table/query/async".format(
+        id=extract_synapse_id_from_query(query)
+    )
+    return QueryResultBundle.fill_from_dict(
+        data=synapse._waitForAsync(uri=uri, request=query_bundle_request)
+    )
+
+
+async def _table_query(
+    query: str, synapse: Optional[Synapse] = None, results_as: str = "csv", **kwargs
+) -> Union["QueryResultBundle", Tuple[Dict[str, str], str]]:
+    """
+    Query a Synapse Table.
+
+    Optional keyword arguments differ for the two return types of `rowset` or `csv`:
+
+    - For `rowset`, you can specify:
+        - `limit`: Maximum number of rows to return.
+        - `offset`: Number of rows to skip before starting to return rows.
+        - `part_mask`: Bit mask to specify which parts of the query result bundle to return.
+
+    - For `csv`, you can specify:
+        - `quote_character`: Character used for quoting fields.
+        - `escape_character`: Character used for escaping quotes within fields.
+        - `line_end`: Character(s) used to terminate lines.
+        - `separator`: Character used to separate fields.
+        - `header`: Whether to include a header row.
+        - `include_row_id_and_row_version`: Whether to include row ID and version in the output.
+        - `download_location`: Location to download the CSV file.
+
+    Returns:
+        If `results_as` is "rowset", returns a QueryResultBundle object.
+        If `results_as` is "csv", returns a tuple of (download_from_table_result, csv_path).
+    """
+    loop = asyncio.get_event_loop()
+    client = Synapse.get_client(synapse_client=synapse)
+
+    if results_as.lower() == "rowset":
+        return query_table_row_set(query=query, synapse=client, **kwargs)
+
+    elif results_as.lower() == "csv":
+        result, csv_path = await loop.run_in_executor(
+            None,
+            lambda: _query_table_csv(
+                query=query,
+                synapse=client,
+                quote_character=kwargs.get("quote_character", DEFAULT_QUOTE_CHARACTER),
+                escape_character=kwargs.get("escape_character", DEFAULT_ESCAPSE_CHAR),
+                line_end=kwargs.get("line_end", str(os.linesep)),
+                separator=kwargs.get("separator", DEFAULT_SEPARATOR),
+                header=kwargs.get("header", True),
+                include_row_id_and_row_version=kwargs.get(
+                    "include_row_id_and_row_version", True
+                ),
+                download_location=kwargs.get("download_location", None),
+            ),
+        )
```

Comment:
Once `_query_table_csv` is made async you will be able to remove the `run_in_executor` logic

Reply:
Mark this as resolved for now since `_query_table_csv` has been turned to async